### PR TITLE
Fix #9344

### LINKED ISSUE #9344: Program built with `--output-complete-exe` cannot read debug infos
When a byte code program needs the debug info, for instance to print a backtrace, the runtime tries to open `Sys.executable_name` as a byte code program to read the debug info. That doesn't work when the program is built with `--output-complete-exe`.

For instance:

```
$ echo "raise Not_found" > test.ml
$ ocamlc -o test -output-complete-exe test.ml
$ ./test
Fatal error: exception Not_found
Fatal error: executable program file not found
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by ghost**
This PR fixes #9344 by preventing programs built with `-output-complete-exe` from trying to open themselves as bytecode files.

To do so, it introduces a global variable `caml_byte_program_mode` that defaults to `STANDARD` and is set to `COMPLETE_EXE` in the `main` function generated for programs built with `-outpuit-complete-exe`.

## Future work

Actually embed the debug info in programs built with `-output-complete-exe`.

### CODE REVIEW

**Comment by ghost** [comment]
@nojb I requested your review as you touched this code recently

**Review by nojb — APPROVED** [review]
LGTM. Thanks!

File: Changes [hunk 1]
```diff
@@ -63,6 +63,10 @@ Working version
 - #9316: Use typing information from Clambda for mutable Cmm variables.
   (Stephen Dolan, review by Vincent Laviron, Guillaume Bury and Xavier Leroy)
 
+- #9344, #9368: Prevent program built with "-output-complete-exe" from trying to
+  open themselves as bytecode file and crashing.
```
Comment (nojb) [inline Changes]: I would insert something like `Disable exception backtraces in bytecode programs built with "-output-complete-exe"` here which may be more informative for the casual user.

Comment (ghost) [reply Changes]: Indeed, I'll update the message

**Comment by ghost** [comment]
Cherry-picked in 4.10: 7e2433c5fc05b4fde05254f5133d5e75fb3e28c0

**Pull request merged**